File: tools/server_observer/observation_api.py
import re
from collections import defaultdict
from dataclasses import dataclass
from datetime import UTC
from datetime import datetime
from datetime import timedelta
from hashlib import sha1
from json import dumps
from time import time

import httpx
from fake_useragent import UserAgent
from httpx import HTTPTransport

from conflict_interface.data_types.authentication import AuthDetails
from conflict_interface.data_types.custom_types import HashMap
from conflict_interface.data_types.custom_types import LinkedList
from conflict_interface.data_types.game_api_types.game_state_action import GameStateAction
from conflict_interface.data_types.game_object_json import dump_any
from conflict_interface.utils.exceptions import AuthenticationException
from conflict_interface.utils.helper import unix_to_datetime
from tools.server_observer.recorder_logger import get_logger

logger = get_logger()
SUPPORTED_CLIENT_VERSION = 207
MAX_RETRIES = 3

@dataclass
class DeviceDetails:
    os: str
    device: str

    @staticmethod
    def from_user_agent(user_agent):
        os = re.findall(r"\(([^;]+);", user_agent)
        if os:
            return DeviceDetails(os[0], "desktop")
        else:
            return DeviceDetails("Unknown", "")


class ObservationApi:
    """
    Handles interactions with the game server API.

    Provides functionality to send authenticated requests, manage proxies,
    fetch game state, calculate client times, and update server time based
    on received data. The class is responsible for maintaining necessary
    details such as game ID, player ID, and server configuration.

    Attributes:
        auth (AuthDetails): Authentication details of the user.
        client_version (int): Client version supported by the server.
        device_details (DeviceDetails): Detailed information about the device,
            derived from the headers.
        proxy (dict): Proxy configuration for the client.
        game_id (int): Identifier for the game being interacted with.
        game_server_address (str): URL of the game server.
        player_id (int): Current player's unique identifier.
        request_id (int): The ID of the current request, starting from 0.
        map_id (int or None): Identifier for the current map (default is None).
        last_update_time (datetime or None): Time of the last server update
            (default is None).
        server_time_offset (timedelta or None): Offset between the server time
            and client time (default is None).

    Methods:
        close(): Close the HTTP client and release resources.
        __enter__(): Support context manager protocol.
        __exit__(): Ensure client is closed when exiting context.
        set_proxy(proxy: dict): Set the proxy configuration.
        unset_proxy(): Clear the current proxy configuration.
        make_game_server_request(parameters): Send a request to the game server
            using the given parameters.
        client_time(time_scale) -> datetime: Calculate the client's adjusted
            time based on a given time scale.
        update_server_time(t_stamp_now): Update the server time offset based
            on the current timestamp.
        get_static_map_data(): Retrieve static map data from the server.
        request_game_state(state_ids: dict, time_stamps: dict) -> tuple[dict,
            dict, dict]: Request and process the game's state based on its IDs
            and timestamps.
    """
    def __init__(self,
                 transport: HTTPTransport,
                 headers: dict,
                 cookies: dict,
                 auth_details: AuthDetails,
                 game_id: int,
                 game_server_address: str,
                 client_version: int = SUPPORTED_CLIENT_VERSION,
                 proxy: dict = None,
                 client: httpx.Client = None):
        # Use provided client if available (for reuse), otherwise create new one
        if client is not None:
            self.client = client
            self._owns_client = False  # Don't close shared client
        else:
            self.client = httpx.Client(transport=transport,
                                       headers=headers,
                                       cookies=cookies,
                                       proxy=proxy.get("http") if proxy else None,
                                       timeout=httpx.Timeout(
                                           connect=10.0,  # seconds to establish connection
                                           read=60.0,  # seconds to wait for a server response
                                           write=30.0,
                                           pool=5.0
                                       ))
            self._owns_client = True
        self.game_id = game_id
        self.player_id = 0
        self.auth = auth_details
        if headers.get("user-agent") is None:
            if headers.get("User-Agent") is not None:
                headers["user-agent"] = headers.get("User-Agent")
            else:
                headers["user-agent"] = UserAgent(platforms='desktop').random
        self.device_details = DeviceDetails.from_user_agent(headers.get("user-agent"))
        self.request_id = 0
        self.client_version = client_version
        self.game_server_address = game_server_address
        self.map_id = None
        self.last_update_time = None
        self.server_time_offset = None
        if proxy:
            self.proxy = proxy
        else:
            self.proxy = defaultdict()

    def close(self):
        """Close the HTTP client and release resources."""
        # Only close if we own the client (not shared)
        if self.client and self._owns_client:
            self.client.close()
            self.client = None

    def __enter__(self):
        """Support context manager protocol."""
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        """Ensure client is closed when exiting context."""
        self.close()
        return False

    def set_proxy(self, proxy: dict):
        self.proxy = proxy

    def unset_proxy(self):
        self.proxy = defaultdict()

    def make_game_server_request(self, parameters):
        attempt = 0

        while True:
            headers = {
                'Accept': 'text/plain, */*; q=0.01',
                'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
                "Accept-Encoding": "gzip, deflate, br"
            }

            hash_hex = sha1(("undefined" + str(int(time() * 1000))).encode()).hexdigest()

            payload = {
                "requestID": self.request_id,
                "language": "en",
                **parameters,
                "version": self.client_version,
                "tstamp": str(self.auth.auth_tstamp),
                "client": "con-client",
                "hash": hash_hex,
                "sessionTstamp": 0,
                "gameID": str(self.game_id),
                "playerID": self.player_id,
                "siteUserID": str(self.auth.user_id),
                "adminLevel": None,
                "rights": self.auth.rights,
                "userAuth": self.auth.auth,
                "lastCallDuration": 0,
            }
            logger.debug(f"Sending Game API request {self.request_id} with params: {dumps(parameters)}")
            self.request_id += 1

            response = self.client.post(self.game_server_address, headers=headers, content=dumps(payload))
            response.raise_for_status()
            response_json = response.json()
            response.close()
            del response

            if not isinstance(response_json.get("result"), int):
                if "timeStamp" in response_json["result"]:
                    self.update_server_time(response_json["result"]["timeStamp"])
            else:
                self.update_server_time(0)

            # Handle errors and recoverable switch
            if isinstance(response_json.get("result"), dict):
                err_class = response_json["result"].get("@c")
                if err_class == "ultshared.UltAuthentificationException":
                    raise AuthenticationException(
                        f"Authentication failed while sending parameters {dumps(payload, indent=2)} to game server.")
                if err_class == "ultshared.rpc.UltSwitchServerException":
                    # Update server address provided by server
                    new_server = "https://" + response_json["result"].get("newHostName")
                    if new_server:
                        logger.info(f"Switching game server to {new_server}")
                        self.game_server_address = new_server

                    # Retry
                    if attempt >= MAX_RETRIES:
                        raise Exception("Exceeded retries after server switch suggestion.")
                    attempt += 1
                    logger.info(f"Retrying after UltSwitchServerException (attempt {attempt}/{MAX_RETRIES})")
                    continue  # re-issue the request
            return response_json

    def client_time(self, time_scale) -> datetime:
        """
        Calculates the client time

        :param time_scale: The time_scale of the game
        """
        current_time = datetime.now(UTC)
        if time_scale not in (0.25, 1, 0.1):
            raise ValueError(f"Time scale cannot be {time_scale}. Must be 0.1, 0.25 or 1")
        if time_scale != 1:
            time_elapsed = timedelta(seconds = (current_time - self.last_update_time).total_seconds() / time_scale)
            return self.last_update_time + self.server_time_offset + time_elapsed
        return current_time + self.server_time_offset

    def update_server_time(self, t_stamp_now):
        self.last_update_time = datetime.now(UTC)

        t_stamp_now = int(t_stamp_now)
        if t_stamp_now == 0:
            seconds_since_epoch = (self.last_update_time - datetime(1970, 1, 1, tzinfo=UTC)).total_seconds()
            self.server_time_offset = timedelta(seconds = -seconds_since_epoch)
            return

        t_stamp_now = unix_to_datetime(t_stamp_now)
        self.server_time_offset = t_stamp_now - self.last_update_time


    def get_static_map_data(self):
        headers = {
            'Accept': 'application/json, text/javascript, */*; q=0.01',
        }

        params = {
            # 'bust': '1700054640135',
        }

        domain = "static1.bytro.com"
        url = f"https://{domain}/fileadmin/mapjson/live/{self.map_id}.json"
        response = self.client.get(
            url,
            params=params,
            headers=headers,
        )

        response.raise_for_status()
        return response.json()

    def request_game_state(self, state_ids: dict, time_stamps: dict) -> tuple[dict, dict, dict]:
        if state_ids and time_stamps:
            include_state_meta = True
            state_ids = HashMap(state_ids)
            time_stamps = HashMap(time_stamps)
        else:
            include_state_meta = False
            state_ids = HashMap()
            time_stamps = HashMap()

        logger.debug(f"Requesting game state with {len(state_ids) if state_ids is not None else 'None'} state IDs and {len(time_stamps) if time_stamps is not None else 'None'} timestamps, including state meta: {include_state_meta}")

        action = GameStateAction(
            state_type=0,
            state_id="0",
            add_state_ids_on_sent=include_state_meta,
            option=None,
            state_ids=state_ids if include_state_meta else None,
            time_stamps=time_stamps if include_state_meta else None,
            actions=LinkedList()
        )

        payload = dump_any(action)
        response = self.make_game_server_request(payload)
        extraction_result = self._extract_state_metadata(response, state_ids, time_stamps)
        if not extraction_result:
            raise Exception(f"Game state extraction failed for {response}")
        return response, state_ids, time_stamps


    def _extract_state_metadata(self, response, state_ids, time_stamps) -> bool:
        """
        Extract state IDs and timestamps from the last raw response
        to enable incremental updates without parsing a GameState.
        """
        result = response.get("result")
        if not isinstance(result, dict):
            return False

        states = result.get("states")
        if not isinstance(states, dict):
            return False

        for state in states.values():
            if not isinstance(state, dict):
                continue
            state_type_raw = state.get("stateType")
            try:
                state_type = int(state_type_raw)
            except (TypeError, ValueError):
                continue

            state_id = state.get("stateID")
            if state_id is not None:
                state_ids[state_type] = str(state_id)

            time_stamp = state.get("timeStamp")
            if time_stamp is not None:
                try:
                    time_stamps[state_type] = int(time_stamp)
                except (TypeError, ValueError):
                    continue

        if len(state_ids) == 0 or len(time_stamps) == 0:
            return False

        return True

File: tools/server_observer/src/observation_api.cpp
#include "observation_api.hpp"
#include <openssl/sha.h>
#include <chrono>
#include <sstream>
#include <iomanip>
#include <simdjson.h>
#include <regex>
#include <utility>


static std::string sha1_hex(const std::string& input) {
    unsigned char hash[SHA_DIGEST_LENGTH];
    SHA1(reinterpret_cast<const unsigned char*>(input.c_str()), input.size(), hash);
    
    std::stringstream ss;
    for (unsigned char i : hash) {
        ss << std::hex << std::setw(2) << std::setfill('0') << static_cast<int>(i);
    }
    return ss.str();
}

static int64_t current_time_ms() {
    auto now = std::chrono::system_clock::now();
    auto ms = std::chrono::duration_cast<std::chrono::milliseconds>(now.time_since_epoch());
    return ms.count();
}

ObservationApi::ObservationApi(
    std::shared_ptr<RequestManager> manager,
    const std::map<std::string, std::string>& headers,
    const std::map<std::string, std::string>& cookies,
    std::shared_ptr<ProxyConfig> proxy,
    AuthDetails  auth_details,
    int game_id,
    const std::string& game_server_address,
    int client_version)
    : game_id_(game_id)
    , player_id_(0)
    , auth_(std::move(auth_details))
    , request_id_(0)
    , client_version_(client_version)
    , game_server_address_(game_server_address)
    , headers_(headers)
    , cookies_(cookies)
    , proxy_(std::move(std::move(proxy)))
{
    cli_ = std::make_unique<HttpClient>(manager, game_server_address);
}

ObservationApi::~ObservationApi() = default;

asio::awaitable<HttpResponse> ObservationApi::request_game_state_async(std::map<std::string, std::string> &state_ids,
                                                                        std::map<std::string, std::string> &time_stamps) {
    bool include_state_meta = !state_ids.empty() && !time_stamps.empty();
    
    // Build state_ids and time_stamps as JSON objects
    json state_ids_json = json::object();
    json time_stamps_json = json::object();
    
    if (include_state_meta) {
        for (const auto& [key, value] : state_ids) {
            state_ids_json[key] = value;
        }
        for (const auto& [key, value] : time_stamps) {
            time_stamps_json[key] = value;
        }
    }
    
    // Build action
    json action = {
        {"@c", "ultshared.action.UltUpdateGameStateAction"},
        {"stateType", 0},
        {"stateID", "0"},
        {"addStateIDsOnSent", include_state_meta}
    };

    if (include_state_meta) {
        action["stateIDs"] = state_ids_json;
        action["stateIDs"]["@c"] = "java.util.HashMap";

        action["timeStamps"] = time_stamps_json;
        action["timeStamps"]["@c"] = "java.util.HashMap";
    }

    action["actions"] = json::array({
        "java.util.LinkedList",
        json::array()
    });
    
    // Configure proxy if provided
    if (proxy_->enabled) {
        cli_->set_proxy(proxy_->host, proxy_->port, proxy_->username, proxy_->password);
    }

    // Build request headers
    Headers req_headers = {
        {"Accept", "text/plain, */*; q=0.01"},
        {"Accept-Encoding", "gzip, deflate, br"}
    };

    // Add custom headers
    for (const auto& [key, value] : headers_) {
        req_headers.insert({key, value});
    }

    // Build payload
    std::string hash_input = "undefined" + std::to_string(current_time_ms());
    std::string hash_hex = sha1_hex(hash_input);

    json payload = {
        {"requestID", request_id_},
        {"language", "en"},
        {"version", client_version_},
        {"tstamp", std::to_string(auth_.auth_tstamp)},
        {"client", "con-client"},
        {"hash", hash_hex},
        {"sessionTstamp", 0},
        {"gameID", std::to_string(game_id_)},
        {"playerID", player_id_},
        {"siteUserID", std::to_string(auth_.user_id)},
        {"adminLevel", nullptr},
        {"rights", auth_.rights},
        {"userAuth", auth_.auth},
        {"lastCallDuration", 0}
    };

    // Merge parameters into payload
    for (auto& [key, value] : action.items()) {
        payload[key] = value;
    }

    request_id_++;

    // Send request asynchronously and return awaitable response
    std::string body = payload.dump();
    co_return co_await cli_->Post_async(req_headers, body, "application/json");
}

GameServerResult ObservationApi::parse_and_validate_response(HttpResponse& response,
                                                             std::map<std::string, std::string> &state_ids,
                                                             std::map<std::string, std::string> &time_stamps) {
    GameServerResult result;
    result.error_code = GameServerError::SUCCESS;
    result.error_message = "";
    
    // Check for timeout first
    if (response.timeout) {
        result.error_code = GameServerError::NETWORK_ERROR;
        result.error_message = response.error_message.empty() ? "Request timed out" : response.error_message;
        return result;
    }
    
    // Check for HTTP errors
    if (response.status_code != 200) {
        result.error_code = GameServerError::HTTP_ERROR;
        result.error_message = "HTTP status: " + std::to_string(response.status_code);
        if (!response.error_message.empty()) {
            result.error_message += " - " + response.error_message;
        }
        return result;
    }

    // Store the raw response string before parsing
    result.raw_response = std::move(response.data);

    // Parse JSON response using simdjson
    simdjson::dom::parser parser;
    simdjson::dom::element game_state;
    try {
        game_state = parser.parse(result.raw_response);
    } catch (const simdjson::simdjson_error& e) {
        result.error_code = GameServerError::PARSE_ERROR;
        result.error_message = "JSON parse error: " + std::string(e.what());
        return result;
    }

    // Validate response structure
    simdjson::dom::object result_obj;
    if (game_state["result"].get(result_obj) != simdjson::SUCCESS) {
        result.error_code = GameServerError::UNKNOWN_ERROR;
        result.error_message = "No result object in response";
        return result;
    }

    std::string_view result_class_view;
    if (result_obj["@c"].get(result_class_view) != simdjson::SUCCESS) {
        result.error_code = GameServerError::UNKNOWN_ERROR;
        result.error_message = "No @c field in result";
        return result;
    }

    std::string result_class(result_class_view);

    // Check for authentication errors
    if (result_class == "ultshared.UltAuthentificationException") {
        result.error_code = GameServerError::AUTH_ERROR;
        result.error_message = "Authentication failed: " + result_class;
        std::string_view message_view;
        if (result_obj["message"].get(message_view) == simdjson::SUCCESS) {
            result.error_message += " - " + std::string(message_view);
        }
        return result;
    }

    // Check for server switch
    if (result_class == "ultshared.rpc.UltSwitchServerException") {
        result.error_code = GameServerError::SERVER_SWITCH;
        result.error_message = "Server switch required";
        std::string_view new_server_view;
        if (result_obj["newHostName"].get(new_server_view) == simdjson::SUCCESS) {
            std::string new_server(new_server_view);
            result.error_message += ": " + new_server;
            cli_->set_url("https://" + new_server);
        }
        return result;
    }

    // Check for client version mismatch
    if (result_class == "ultshared.UltClientVersionMismatchException") {
        result.error_code = GameServerError::CLIENT_VERSION_MISMATCH;
        result.error_message = "Client version mismatch";

        std::string_view message_view;
        if (result_obj["detailMessage"].get(message_view) == simdjson::SUCCESS) {
            int old_client_version = -1;
            int new_client_version = -1;
            std::string message(message_view); // convert string_view to string for regex

            std::regex re1("con-client: #(\\d+)");
            std::regex re2("con-client_live\\.txt: #(\\d+)");

            std::smatch match1, match2;

            if (std::regex_search(message, match1, re1)) {
                old_client_version = std::stoi(match1[1]);
            }

            if (std::regex_search(message, match2, re2)) { // use 'message' here
                new_client_version = std::stoi(match2[1]);
            }
            if (old_client_version != -1 && new_client_version != -1) {
                result.error_message += ": old version " + std::to_string(old_client_version) +
                                        ", new version " + std::to_string(new_client_version);
                client_version_ = new_client_version;
            }
            else {
                result.error_message += ": " + message;
            }
        }
    }
    // Check for valid game state
    if (result_class != "ultshared.UltAutoGameState" && result_class != "ultshared.UltGameState") {
        result.error_code = GameServerError::UNKNOWN_ERROR;
        result.error_message = "Unknown error class: " + result_class;
        std::string_view message_view;
        if (result_obj["message"].get(message_view) == simdjson::SUCCESS) {
            result.error_message += " - " + std::string(message_view);
        }
        return result;
    }

    // Extract state metadata using simdjson
    simdjson::dom::object states_obj;
    if (result_obj["states"].get(states_obj) != simdjson::SUCCESS) {
        result.error_code = GameServerError::UNKNOWN_ERROR;
        result.error_message = "Game state extraction failed";
        return result;
    }

    // Process states to extract metadata
    bool game_ended = false;
    for (auto [key, state_val] : states_obj) {
        simdjson::dom::object state;
        if (state_val.get(state) != simdjson::SUCCESS) {
            continue;
        }

        // Check if state has stateType field
        uint64_t state_type;
        if (state["stateType"].get(state_type) != simdjson::SUCCESS) {
            continue;
        }

        std::string_view state_id_view;
        if (state["stateID"].get(state_id_view) == simdjson::SUCCESS) {
            state_ids[std::string(key)] = std::string(state_id_view);
        }

        std::string_view timestamp_view;
        if (state["timeStamp"].get(timestamp_view) == simdjson::SUCCESS) {
            time_stamps[std::string(key)] = std::string(timestamp_view);
        }

        // Check if game has ended
        bool ended;
        if (state["gameEnded"].get(ended) == simdjson::SUCCESS && ended) {
            game_ended = true;
        }
    }

    if (state_ids.empty() || time_stamps.empty()) {
        result.error_code = GameServerError::UNKNOWN_ERROR;
        result.error_message = "Game state extraction failed";
        return result;
    }

    // Success - return with minimal JSON parsing
    result.error_code = GameServerError::SUCCESS;
    result.error_message = "";
    result.game_ended = game_ended;
    // No need to parse the full JSON for data field in the success case
    return result;
}

json ObservationApi::get_static_map_data(int map_id) {
    std::string url = "https://static1.bytro.com/fileadmin/mapjson/live/" +
                      std::to_string(map_id) + ".json";
    httplib::Client cli("https://static1.bytro.com");
    cli.set_follow_location(true);

    // Configure proxy if provided
    if (proxy_->enabled) {
        cli.set_proxy(proxy_->host, proxy_->port);

        // Set proxy authentication if provided
        if (!proxy_->username.empty() && !proxy_->password.empty()) {
            cli.set_proxy_basic_auth(proxy_->username, proxy_->password);
        }
    }

    httplib::Headers headers = {
        {"Accept", "application/json, text/javascript, */*; q=0.01"}
    };

    std::string path = "/fileadmin/mapjson/live/" + std::to_string(map_id) + ".json";
    auto res = cli.Get(path, headers);

    if (!res || res->status != 200) {
        throw std::runtime_error("Failed to fetch static map data");
    }

    // Parse and immediately clear the response body to free memory
    json result = json::parse(res->body);
    res->body.clear();
    res->body.shrink_to_fit();
    return result;
}

void ObservationApi::update_package(ObservationPackage &pkg) const {
    pkg.client_version = client_version_;
    pkg.game_server_address = game_server_address_;
    pkg.headers = headers_;
    pkg.cookies = cookies_;
    pkg.auth = auth_;
}